### PR TITLE
Fix GDSCript inner classes be freed after compile

### LINKED ISSUE #29395: Cannot use inner classes of name classes  as type decleration
**Godot version:**

Godot 3.2-dev compiled from 449395716f8bacafc667ce1cd5cd8e35572db0b2

**OS/device including version:**

Arch Linux

**Issue description:**

GDScript parse failed when use inner classes of named script as type declearation. The inner class cannot be resoved.

Godot 3.1.1 works fine.

**Steps to reproduce:**

ValueEvaluator.gd
```gdscript
tool
extends EventEmitter
class_name ValueEvaluator

class IEvolutor:
	func has_value(prop) -> bool:
		return false
	func get_value(prop):
		return null

class Register extends IEvolutor:
	var values = {}
	func has_value(prop) -> bool:
		return prop in values
	func get_value(prop):
		return values[prop]
```

Condition.gd
```gdscript
tool
extends EventEmitter
class_name Condition

var register: ValueEvaluator.Register = null
```

The compiler compains as below
```
SCRIPT ERROR: GDScript::reload: Parse Error: Identifier 'Register' is not a valid type (not a script or class), or could not be found on base 'ValueEvaluator'.
   At: res://addons/simple-game-framework/core/condition/Condition.gd:5.

```

### LINKED ISSUE #29396: Cannot use constants of script preloaded in statck
**Godot version:**

Godot 3.2-dev compiled from 449395716f8bacafc667ce1cd5cd8e35572db0b2
Godot 3.1.1 works as expected.

**OS/device including version:**

Arch Linux

**Issue description:**

Resources got freed too early in the statck.
We should avoid preloaded resources be freed before the script itself is freed.

**Steps to reproduce:**

```gdscript
class IEvolutor:
	func has_value(prop) -> bool:
		return false
	func get_value(prop):
		return null

class Register extends IEvolutor:
	var values = {}
	func has_value(prop) -> bool:
		return prop in values
	func get_value(prop):
		return values[prop]

```

```gdscript
extends Control

var RegisterClass = null
var register_instance = null

func _ready():
	RegisterClass = preload("ValueEvaluator.gd").Register
	register_instance = RegisterClass.new()
```

Got a runtime error as below:
```
Attempt to call function 'new' in base 'previously freed instance' on a null instance.
```

### PR DESCRIPTION
**Pull request by Geequlim**
Introduce in #27128
Fix #29395 
Fix #29396

### CODE REVIEW

File: modules/gdscript/gdscript_compiler.cpp [hunk 1]
```diff
@@ -1977,12 +1977,12 @@ Error GDScriptCompiler::_parse_class_level(GDScript *p_script, const GDScriptPar
 
 	for (int i = 0; i < p_class->subclasses.size(); i++) {
 		StringName name = p_class->subclasses[i]->name;
-
-		GDScript *subclass = p_script->subclasses[name].ptr();
+		Ref<GDScript>& subclass = p_script->subclasses[name];
```
Comment (akien-mga) [inline modules/gdscript/gdscript_compiler.cpp]: Style issue here, it should be `Ref<GDScript> &subclass`.

**Comment by akien-mga** [comment]
CC @bojidar-bg

**Comment by bojidar-bg** [comment]
Looks good, but doesn't the `p_script->subclasses` map already the reference alive?

**Comment by Geequlim** [comment]
@bojidar-bg  'p_script' might be freed.

**Review by bojidar-bg — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**